Docs: add more informaton to Architecture page

Luanf opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
Update Architecture docs page with more information regarding how the order of test execution is decided and how Jest handles parallelization, as pointed in #6957.

Note: the issue was more directed to just the parallelization part, but I thought it made sense to add the test ordering information as well.
Please let me know if it would be better to move that information elsewhere or not including it at all.
It's my first contribution here so I'm very open to guidance. :smile:

Luanf commented:
I did accept the CLA a few minutes before opening this PR so maybe it's taking a bit to update...

Luanf commented:
Unsure if I should add the docs to the _/versioned_docs/version-24.1/_. Thoughts?

Luanf commented:
Ops. I see some linting errors on my .md changes.
I'm fixing it

SimenB commented:
Thank you!

>Unsure if I should add the docs to the _/versioned_docs/version-24.1/_

It would be nice if the architecture page was not versioned - it's not API, so it'd not necessarily tied to any version.

That said - for now please also update the versioned docs 🙂

rickhanlonii commented:
@Luanf if you can fix the conflict we can merge

rickhanlonii commented:
Hey @Luanf I just re-read the page and doesn't this seem a bit empty or out of place:

![deploy-preview-7984--jest-preview netlify com_docs_en_architecture](https://user-images.githubusercontent.com/2440089/53407807-edeadf00-39b4-11e9-82f7-49f251326995.png)

Is there a better place to put this content? Maybe in a code comment?

Luanf commented:
Thanks for the review @rickhanlonii !
I agree it doesn't look like a perfect fit :sweat_smile: 
I'm not entirely sure what is the intended scope for the Architecture page, tbh
I would be happy to move it to a more fitting place, but I'm not sure what you mean by code comment.
Can you elaborate please?

cpojer commented:
Could we instead put it into the code as comments where this is most relevant?

Luanf commented:
As an user, I would like to be able to read somewhere on the online docs and find out more information about ordering & parallelization, and I think that was the reason #6957 got created (although it was only about parallelization, I thought ordering could go well with that info).

As for adding code comments, I went quickly through the code and the ordering part seems reasonably well documented in _TestSequencer.js_. And although the parallelization part seems to be more scattered out, the info about the runInBand behavior is commented on _testSchedulerHelper.js_.
So, if it's about adding more code comments I'm not sure where would they be and I could use some suggestions.

But still I personally would think this info belongs to somewhere in the official docs.

cpojer commented:
Check out this long comment: https://github.com/facebook/jest/blob/master/packages/jest-haste-map/src/index.ts#L147

I think we can do something similar here, and link to it from some other places.

Luanf commented:
I had more time to take a look at this one so I went through the code and attempted to set out some updates to code comments for the Test Sequencer and the Test Scheduler.

I've also removed the notes I had added to the website docs, although that was the point asked on #6957.

Thanks for all your reviews so far and please let me know if there is something more I could do here :smile:

cpojer commented:
Thank you!

akuji1993 commented:
I had to dig through a lot of Google results to get to this issue. Keeping that in mind, I really don't think the decision to pull the documentation from the site and keep it in the code was the right one. This way I get greeted by an empty "Architecture" page, that just has a link to a 50 minute YouTube video instead of some structured text that I can fly over and find what I need. 

The text that @Luanf provided here (https://github.com/facebook/jest/pull/7984#issuecomment-467393723) was perfect, it was exactly what I was looking for and what a colleague was looking for just two weeks ago. I'd really recommend not having documentation in code only, where nobody will find it that reads the official documentation website.

cpojer commented:
@akuji1993 Please feel free to send pull requests to add an architecture overview to the Jest docs.
